lib/Optimizer/GraphOptimizer.cpp

```diff
@@ -885,34 +885,52 @@ static void optimizePool(Function *F) {
   } // For all nodes in the graph.
 }
 
-/// \returns The uniquely used Constant from node or nullptr
-/// if node has more than one user or is not a Constant.
-static Constant *getUniquelyUsedConstant(Node &node) {
-  // If that node has more than one use, it may not
-  // be okay to modify the underlying Constant.
-  if (!node.hasOneUse()) {
+/// \returns a uniquely used Constant with the same contents as \p node. If \p
+/// node is not a Constant then \returns a nullptr. If \node is a Constant which
+/// has a single use, \p node is returned. If \node is a Constant which has
+/// multiple uses, then \returns a new duplicate Constant that has the same
+/// contents as \p node.
+static Constant *getUniquelyUsedConstant(Function *F, Node &node) {
+  Constant *constant = dyn_cast<Constant>(&node);
+  if (!constant) {
     return nullptr;
   }
-  return dyn_cast<Constant>(&node);
+
+  if (constant->hasOneUse()) {
+    return constant;
+  }
+
+  // If constant has more than one use, duplicate it and return the duplicate.
+  auto *NC =
+      F->getParent()->createConstant(constant->getType(), constant->getName());
+  NC->getPayload().assign(&constant->getPayload());
+  return NC;
 }
 
-/// Normalize the weight of \p CV with what \p BN is doing.
-/// \return Whether or not the normalization was possible.
+/// Normalize the weight of \p CV with what \p BN is doing.  If normalization is
+/// possible, \returns a ConvolutionNode with the same inputs as \p CV but with
+/// normalized filter and bias; otherwise \returns a nullptr.
 template <typename ElemTy>
-bool normalizeWeights(ConvolutionNode &CV, BatchNormalizationNode &BN) {
-
+ConvolutionNode *normalizeWeights(Function *F, ConvolutionNode &CV,
+                                  BatchNormalizationNode &BN) {
   static_assert(
       std::is_floating_point<ElemTy>::value ||
           std::is_same<float16_t, typename std::remove_cv<ElemTy>::type>::value,
       "This implementation is for floating-point values only");
 
-  Constant *filterC = getUniquelyUsedConstant(*CV.getFilter().getNode());
-  Constant *cbiasC = getUniquelyUsedConstant(*CV.getBias().getNode());
+  Constant *filterC = getUniquelyUsedConstant(F, *CV.getFilter().getNode());
+  Constant *cbiasC = getUniquelyUsedConstant(F, *CV.getBias().getNode());
 
   if (!filterC || !cbiasC) {
-    return false;
+    return nullptr;
   }
 
+  // Create a new convolution node with the same parameters and input as CV, but
+  // with unique filter/bias which will be normalized below.
+  auto *newCV = F->createConv(CV.getName(), CV.getInput(), filterC, cbiasC,
```

Comment:
Maybe add a comment that we need to create a new node because the constants may not be bound to that node.

Alternatively, we could always call setNThInput on the related arguments. BTW it may be better because it would avoid to rely on DCE to get rid of the now unconnected convolution node.